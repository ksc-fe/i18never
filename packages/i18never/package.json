{
  "name": "i18never",
  "version": "0.0.1",
  "description": "Deal with the internationalization (i18n) automatically.",
  "main": "index.js",
  "scripts": {
    "test": "jest"
  },
  "keywords": [
    "i18n"
  ],
  "author": "Javey",
  "license": "MIT",
  "dependencies": {
    "@babel/generator": "^7.17.7",
    "@babel/parser": "^7.17.8",
    "@babel/traverse": "^7.17.3",
    "@babel/types": "^7.17.0",
    "@inquirer/rawlist": "^0.0.7-alpha.0",
    "graphql": "15",
    "graphql-request": "^4.3.0",
    "inquirer": "^8.2.4"
  }
}
