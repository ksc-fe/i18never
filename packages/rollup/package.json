{
    "name": "rollup-plugin-i18never",
    "version": "1.0.24",
    "description": "A rollup plugin based on i18never",
    "main": "dist/index.js",
    "module": "dist/index.esm.js",
    "types": "dist/index.d.ts",
    "scripts": {
        "test": "jest"
    },
    "keywords": [
        "i18never",
        "i18n",
        "rollup-plugin"
    ],
    "publishConfig": {
        "access": "public",
        "registry": "https://registry.npmjs.org/"
    },
    "repository": {
        "type": "git",
        "url": "git+https://github.com/ksc-fe/i18never.git"
    },
    "bugs": {
        "url": "https://github.com/ksc-fe/i18never/issues"
    },
    "author": "Javey",
    "license": "MIT",
    "devDependencies": {
        "rollup": "^2.79.0"
    },
    "dependencies": {
        "@i18never/shared": "^1.0.23",
        "@i18never/transform": "^1.0.23"
    },
    "gitHead": "9cf17c1522768da3ae6727412f750b74e2508371"
}
