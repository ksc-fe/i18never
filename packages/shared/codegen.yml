overwrite: true
schema:
  - http://i18never.ksyun.com/graphql:
      headers:
        Authorization: "${I18NEVER_TOKEN}"
generates:
  src/graphql/index.ts:
    documents: 'src/*.gql'
    plugins:
      - "typescript"
      # - "typescript-resolvers"
      - "typescript-operations"
      - "typescript-graphql-request"
