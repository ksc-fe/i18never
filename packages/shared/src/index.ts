export * from './graphql';
export * from './visitors';
export * from './helpers';
export * from './options';
export * from './plugin';
export { getSdk } from './helpers';
