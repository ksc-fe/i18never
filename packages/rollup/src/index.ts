import { TransformResult } from 'rollup';
import { i18nTrans } from '@i18never/parser';
import { getSdk } from '@i18never/graphql';
import { GraphQLClient } from 'graphql-request';
import { TempKeyItem } from '@i18never/parse/dist/types';
import { resolve, join } from 'path';
import { existsSync } from 'fs';

interface Options {
    root?: string;
    langKey?: string;
    storageType?: string;
    token?: string;
}

enum EnforceType {
    PRE = 'pre',
    POST = 'post',
}

export default function i18never(options: Options = {}) {
    const allAppKeys: TempKeyItem[] = [];
    let i18nVersion = '';
    const { root = 'src' } = options;
    const rootPath = resolve(process.cwd(), root);
    return {
        name: 'i18never',
        enforce: EnforceType.POST,
        options(inputOptions) {
            const configPath = join(process.cwd(), '.i18neverrc.js')
            if (!existsSync(configPath) && !options.token) {
                throw new Error('Missing permission information. Please configure the. i18neverrc.js file in the root directory.');
            }

            if (options.token) {
                process.env.I18NEVER_TOKEN = options.token;
            } else {
                const configOptions = loadConfigFile(configPath);
                process.env.I18NEVER_TOKEN = configOptions.token;
            }
            return inputOptions;
        },
        async transform(code: string, id: string) {
            if (
                !id.match(/\.(pug|vue|tsx|jsx|js|ts)$/) ||
                id.includes('@i18never/client') ||
                !id.startsWith(rootPath)
            ) {
                return;
            }
            const { transCode, allKeys } = await i18nTrans(code, id);
            allAppKeys.push(...allKeys);
            return { code: transCode, map: null } as
                | TransformResult
                | Promise<TransformResult>;
        },
        async buildEnd() {
            const version = await queryVersion(allAppKeys);
            console.log(version);
            i18nVersion = version;
        },
        async generateBundle(_, bundle) {
            const entryHtmlFile = bundle['index.html'];
            if (entryHtmlFile) {
                const html = entryHtmlFile.source;

                const newHtml = html.replace(
                    '</head>',
                    `${generateScript(i18nVersion, options)}</head>`
                );
                entryHtmlFile.source = newHtml;
            }
        },
    };
}

function generateScript(version, options: Options) {
    let lang = '';
    const { langKey = '', storageType = '' } = options;
    switch (storageType) {
        case 'cookie':
            lang = `document.cookie.replace(/(?:(?:^|.*;\\s*)${langKey}\\s*\=\\s*([^;]*).*$)|^.*$/, "$1");`;
            break;
        case 'localStorage':
        case 'sessionStorage':
            lang = `${storageType}.getItem('${langKey}')`;
            break;
        default:
            lang = `document.cookie.replace(/(?:(?:^|.*;\\s*)ksc_lang\\s*\=\\s*([^;]*).*$)|^.*$/, "$1");`;
            break;
    }
    return `
    <script>
        let lang = ${lang};
        document.write('<scr'+'ipt src="http://i18never.ksyun.com/dict/'+lang+'/${version}"></scr'+'ipt>');
    </script>
    `;
}

async function queryVersion(allAppKeys) {
    const client = new GraphQLClient('http://i18never.ksyun.com/graphql/', {
        headers: {
            Authorization: process.env.I18NEVER_TOKEN as string,
        },
    });
    const sdk = getSdk(client);

    const { getVerionId: data } = await sdk.CreateVersion({
        source: 'i18never',
        values: _sortKeys(allAppKeys),
    });
    return data.Id;
}

function loadConfigFile(filePath) {
    const config = require(filePath);
    return config;
}

function _sortKeys(allAppKeys) {
    const targetValues = allAppKeys.map(({ key, tags }) => {
        return {
            key,
            tags: tags
                ? Object.keys(tags).map((language) => {
                      const name = tags[language];
                      return { language, name };
                  })
                : undefined,
        };
    });

    targetValues.sort((a, b) => {
        a.tags.sort((tagA, tagB) => {
            if (tagA.language < tagB.language) {
                return -1;
            } else if (tagA.language > tagB.language) {
                return 1;
            } else {
                return 0;
            }
        });

        b.tags.sort((tagA, tagB) => {
            if (tagA.language < tagB.language) {
                return -1;
            } else if (tagA.language > tagB.language) {
                return 1;
            } else {
                return 0;
            }
        });

        if (a.key < b.key) {
            return -1;
        } else if (a.key > b.key) {
            return 1;
        } else {
            return 0;
        }
    });

    return targetValues;
}
