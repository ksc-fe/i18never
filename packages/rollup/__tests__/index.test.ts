import i18never from '../src';
import { RollupOptions, rollup } from 'rollup';
import path from 'path';
import { CreateVersionQuery } from '@i18never/shared';

function MockGraphQLClient() {
    // emtpy
}
MockGraphQLClient.prototype.request = function (): Promise<CreateVersionQuery> {
    return Promise.resolve({
        getVerionId: {
            Id: 'fake_id',
        },
    });
};
jest.mock('graphql-request', () => {
    return {
        GraphQLClient: MockGraphQLClient,
    };
});

test('should generate file correctly', async () => {
    const config: RollupOptions = {
        input: path.resolve(__dirname, './assets/entry.js'),
        plugins: [i18never()],
    };
    const buddle = await rollup(config);

    await buddle.write({
        file: path.resolve(__dirname, './dist/buddle.js'),
        format: 'es',
        interop: 'esModule',
    });
});
