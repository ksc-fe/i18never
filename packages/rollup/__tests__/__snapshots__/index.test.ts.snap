// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate file correctly 1`] = `
"import { $_ } from '@i18never/client';

const a = $_(\\"测试\\");
const b = \\"测试\\";
const c = $_(\\"购买{0}台\\", [a], { \\"jp\\": \\"v\\" });

console.log(a, b, c);
"
`;

exports[`should generate file correctly 2`] = `
"
<!doctype html>
<html lang=\\"en\\">
  <head>
    <meta charset=\\"utf-8\\">
    <title>Rollup Bundle</title>
    
  <script>
        var lang = ('; '+document.cookie).split('; ksc_lang=').pop().split(';').shift() || 'zh';
        document.write('<scr'+'ipt src=\\"//i18never.ksyun.com/dict/'+lang+'/fake_id\\"></scr'+'ipt>');
    </script></head>
  <body>
    <script src=\\"bundle.js\\" type=\\"module\\"></script>
  </body>
</html>"
`;
